assembly/src/assembler/instruction/field_ops.rs

```diff
@@ -95,22 +95,144 @@ pub fn pow2(span: &mut SpanBuilder) -> Result<Option<CodeBlock>, AssemblyError>
     Ok(None)
 }
 
+/// Computes `ilog2` for the number at the top of the stack, and its corresponding power-of-two.
```

Comment:
We could use the same format of [rust documentation](https://doc.rust-lang.org/std/primitive.u64.html#method.ilog2) to describe the functionality (that is "Returns the base 2 logarithm of the number, rounded down.").

I think we need to elaborate more on the meaning of `pow2` as well. If I understand the implementation correctly, it is the resulting ilog to the power of `2`. It could, then, look like this

```
/// Input: [n, ...]
/// Output: [2^log2(n), log2(n), ...]
```

Maybe we add also an explanation on why we preserve `2^log2(n)`? The other arithmetic operations usually takes two operands are returns a number. We are making an exception here to also return the auxiliary value; it might be good if we can clearly see in what cases we should return such number - if we don't have that many cases, we will force the user to always drop this number before getting the desired result.

One more remark: `0` is an invalid input  for this function. We might want to highlight that here.

Finally, regarding the naming, this is more of a neutral point but `ilog2` makes sense on the context of Rust - where we have floating pointers. We don't have floating pointers here, so the `i` prefix has no purpose there (to differentiate with logarithm that will not round down). OTOH, it might be a good thing to keep synergy with Rust (hence, keep the same method name).

Reply:
@vlopes11 sorry, maybe i should have made this a draft PR and not requested review yet. I introduced the `ilog2` in another PR, mentioned on this PR's text (for reference https://github.com/0xPolygonMiden/miden-vm/pull/868). I guess I can just close that one and we proceed with this since you already had the work to review the two changes together.

Reply:
> We could use the same format of [rust documentation](https://doc.rust-lang.org/std/primitive.u64.html#method.ilog2) to describe the functionality (that is "Returns the base 2 logarithm of the number, rounded down.").

Changed

> I think we need to elaborate more on the meaning of pow2 as well. If I understand the implementation correctly, it is the resulting ilog to the power of 2. It could, then, look like this

Changed to `Output: [2^ilog2, ilog2, ...]`

> Maybe we add also an explanation on why we preserve 2^log2(n)? The other arithmetic operations usually takes two operands are returns a number. We are making an exception here to also return the auxiliary value; it might be good if we can clearly see in what cases we should return such number - if we don't have that many cases, we will force the user to always drop this number before getting the desired result.

Copying my comment from the other PR:

> The instruction returns the `pow2` because it needs to compute it, and this value is useful for other operations, e.g. the MMR uses that.
> 
> But that means the name of the instruction isn't great, I couldn't think of a better alternative :(

https://github.com/0xPolygonMiden/miden-vm/pull/868#discussion_r1170633222

> One more remark: 0 is an invalid input for this function. We might want to highlight that here.

added

> Finally, regarding the naming, this is more of a neutral point but ilog2 makes sense on the context of Rust - where we have floating pointers. We don't have floating pointers here, so the i prefix has no purpose there (to differentiate with logarithm that will not round down). OTOH, it might be a good thing to keep synergy with Rust (hence, keep the same method name).

which would should be used?